Mathlib/Algebra/Group/Action/Equidecomp.lean

```diff
@@ -0,0 +1,231 @@
+
+More familiarly, the restriction of `f` to `A` is the result of partitioning `A` into finitely many
+pieces, then applying a single element of `G` to each piece. -/
+def IsDecompOn (f : X → X) (A : Set X) (S : Finset G) : Prop := ∀ a ∈ A, ∃ g ∈ S, f a = g • a
+
+variable (X G)
+
+/-- Let `G` act on a space `X`. An `Equidecomposition` with respect to `X` and `G` is a partial
+bijection `f : PartialEquiv X X` with the property that for some set `elements : Finset G`,
+(which we record), for each `a ∈ f.source`, `f a` can be obtained by applying some `g ∈ elements`
+instead. We call `f` an equidecomposition of `f.source` with `f.target`.
+
+More familiarly, `f` is the result of partitioning `f.source` into finitely many pieces,
+then applying a single element of `G` to each to get a partition of `f.target`.
+-/
+structure _root_.Equidecomp extends PartialEquiv X X where
+  decomp' : ∃ S : Finset G, IsDecompOn toFun source S
+
+variable {X G}
+
+/-- Note that `Equidecomp X G` is not `FunLike`. -/
+instance : CoeFun (Equidecomp X G) fun _ => X → X := ⟨fun f => f.toFun⟩
+
+/-- A finite set of group elements witnessing that `f` is an equidecomposition. -/
+noncomputable
+def elements (f : Equidecomp X G) := f.decomp'.choose
+
+theorem decomp (f : Equidecomp X G) : IsDecompOn f f.source f.elements := f.decomp'.choose_spec
+
+@[simp]
+theorem map_source {f : Equidecomp X G} {x : X} (h : x ∈ f.source) :
+    f x ∈ f.target := f.toPartialEquiv.map_source h
+
+theorem toPartialEquiv_injective : Injective <| toPartialEquiv (X := X) (G := G) := by
+  intro ⟨_, _, _⟩ _ _
+  congr
+
+theorem IsDecompOn.mono {f f' : X → X} {A A' : Set X} {S : Finset G} (h : IsDecompOn f A S)
+    (hA' : A' ⊆ A) (hf' : EqOn f f' A') : IsDecompOn f' A' S := by
+  intro a ha
+  rw [← hf' ha]
+  exact h a (hA' ha)
+
+/-- The restriction of an equidecomposition as an equidecomposition. -/
+@[simps!]
+def restr (f : Equidecomp X G) (A : Set X) : Equidecomp X G where
+  toPartialEquiv := f.toPartialEquiv.restr A
+  decomp' := ⟨f.elements,
+    f.decomp.mono (source_restr_subset_source _ _) fun _ ↦ congrFun rfl⟩
+
+@[simp]
+theorem restr_toPartialEquiv (f : Equidecomp X G) (A : Set X) :
+    (f.restr A).toPartialEquiv = f.toPartialEquiv.restr A := rfl
+
+theorem restr_source_of_subset (f : Equidecomp X G) {A : Set X} (hA : A ⊆ f.source) :
+    (f.restr A).source = A := by rw [restr_source, inter_eq_self_of_subset_right hA]
+
+theorem restr_eq_of_source_subset {f : Equidecomp X G} {A : Set X} (hA : f.source ⊆ A) :
+    f.restr A = f := by
+  apply toPartialEquiv_injective
+  rw [restr_toPartialEquiv, PartialEquiv.restr_eq_of_source_subset hA]
+
+@[simp]
+theorem restr_univ (f : Equidecomp X G) : f.restr univ = f :=
+  restr_eq_of_source_subset <| subset_univ _
+
+end SMul
+
+open scoped Classical
+
+section Monoid
+
+variable [Monoid G] [MulAction G X]
+
+variable (X G)
+
+/-- The identity function is an equidecomposition of the space with itself. -/
+def refl : Equidecomp X G where
```

Comment:
```suggestion+
@[simps]
def refl : Equidecomp X G where
```

Reply:
similarly, should this be `simps toPartialEquiv`